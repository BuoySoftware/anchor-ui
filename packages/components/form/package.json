{
  "name": "@buoysoftware/anchor-form",
  "version": "0.14.0-alpha.0",
  "main": "src/index.ts",
  "license": "MIT",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc -outDir dist",
    "compile": "tsc --noEmit",
    "test": "jest",
    "postpack": "clean-package restore",
    "prepack": "clean-package"
  },
  "dependencies": {
    "@buoysoftware/anchor-layout": "^0.14.0-alpha.0",
    "@buoysoftware/anchor-typography": "^0.14.0-alpha.0",
    "i18next": "^22.4.9",
    "lodash": "*",
    "react-hook-form": "^7.42.1",
    "react-i18next": "^12.1.4"
  },
  "devDependencies": {
    "jest": "^29.3.1",
    "ts-jest": "^29.0.4",
    "ts-node": "^10.9.1"
  }
}
