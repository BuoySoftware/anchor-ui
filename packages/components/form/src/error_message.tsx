import { WarningCircle } from "iconoir-react";
import { Box, BoxProps, Flex } from "@buoysoftware/anchor-layout";
import { Body, Markdown } from "@buoysoftware/anchor-typography";
import { useTheme } from "@buoysoftware/anchor-theme";

interface OwnProps {
  errors: string[];
  parentId: string;
}

type ErrorMessageProps = OwnProps & BoxProps;

export const ErrorMessage: React.FC<ErrorMessageProps> = ({
  errors,
  parentId,
  ...props
}): React.ReactElement => {
  const { colors } = useTheme();
  const id = `error-${parentId}`;

  return (
    <Flex
      bg="background.criticalSubdued"
      borderRadius="4px"
      data-testid={id}
      id={id}
      py="xxs"
      px="xs"
      role="alert"
      {...props}
    >
      <WarningCircle color={colors.text.critical} width={16} height={16} />
      <Box ml="xs">
        {errors.map((error, index) => (
          <Body as="div" size="m" key={`error-${index}`} color="text.critical">
            <Markdown>{error}</Markdown>
          </Body>
        ))}
      </Box>
    </Flex>
  );
};
