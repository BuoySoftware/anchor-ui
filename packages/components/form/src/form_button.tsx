import { useFormState } from "react-hook-form";
import { Button, ButtonProps } from "@buoysoftware/anchor-button";

type FormButtonProps = ButtonProps;

export const FormButton: React.FC<FormButtonProps> = ({
  children,
  ...props
}): React.ReactElement => {
  const { isSubmitting } = useFormState();

  return (
    <Button submitting={isSubmitting} {...props}>
      {children}
    </Button>
  );
};
