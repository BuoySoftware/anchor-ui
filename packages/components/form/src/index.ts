export * from "./checkbox";
export * from "./checkbox_field";
export * from "./error_message";
export * from "./form";
export * from "./form_base_errors";
export * from "./form_button";
export * from "./form_scope_provider";
export * from "./input";
export * from "./label";
export * from "./radio_field";
export * from "./select";
export * from "./select_field";
export * from "./text_area";
export * from "./text_field";
export * from "./use_conditional_field";
export * from "./use_form_field";

export { default as anchorFormsEn } from "./i18n/anchorForms.en.json";
