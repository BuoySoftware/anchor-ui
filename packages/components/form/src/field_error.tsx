import compact from "lodash/compact";
import flatten from "lodash/flatten";
import { FieldError as HookFormFieldError } from "react-hook-form";
import snakeCase from "lodash/snakeCase";
import { BoxProps } from "@buoysoftware/anchor-layout";
import { TFunction } from "i18next";
import { useTranslation } from "react-i18next";

import { useFormScope } from "./form_scope_provider";
import { ErrorMessage } from "./error_message";

export type InputType =
  | "checkbox"
  | "collectionCheckbox"
  | "hidden"
  | "phoneNumber"
  | "radio"
  | "select"
  | "text";

interface OwnProps {
  error?: HookFormFieldError;
  name?: string;
  inputId: string;
  inputType?: InputType;
}

export type FieldErrorProps = OwnProps & BoxProps;

/**
 * Look up an error message in the i18next translations. This utilizes the
 * i18next fallback mechanism for determining who to translate the error message.
 *
 * The order is as follows:
 *
 * 1. An error message in a custom provided namespace via tNamespace scoped to
 *    forms.errors and then the form scope, field name, and error typea.
 *
 * 2. An error message in the default forms namespace scoped to the form scope,
 *    field name and error type.
 *
 * 3. An error message in default forms namespace for the given input type and
 *    error type.
 *
 * 4. An error message in the anchorForms namespace for the given input type and
 *    error type.
 *
 * 5. An error message in the forms namespace for the given error type
 *
 * 6. An error message in the anchorForms namespace for the given error type
 */
function lookupErrorMessage(
  error: HookFormFieldError,
  scope: string,
  name: string,
  inputType: InputType,
  t: TFunction,
  tNamespace: string[]
): string {
  if (error.message !== undefined && error.message !== "") {
    return error.message;
  }

  const i18nName = snakeCase(name);
  const i18nErrorScope = snakeCase(scope);

  // These are translations scoped to the form's scope and field name.
  // The use of forms.errors is to look into the custom tNamespace if one was
  // provided. The second will look up in the forms default namespace.
  const fieldScopes = [
    `forms.errors.${i18nErrorScope}.${i18nName}.${error.type}`,
    `errors.${i18nErrorScope}.${i18nName}.${error.type}`,
  ];

  // Global scopes are used to look up defaults in the forms namespace or
  // the anchorForms namespace
  const globalScopes = [
    `errors.input_types.${snakeCase(inputType)}.${error.type}`,
    `errors.${error.type}`,
  ];

  const combinedGlobalScopes = tNamespace.flatMap((namespace: string) =>
    globalScopes.map((scope) => `${namespace}:${scope}`)
  );

  return t([...fieldScopes, ...combinedGlobalScopes]);
}

export const FieldError: React.FC<FieldErrorProps> = ({
  error,
  inputId,
  name,
  inputType = "text",
  ...elementProps
}): React.ReactElement | null => {
  const { scope, tNamespace = [] } = useFormScope();
  const namespace = compact(flatten([tNamespace, "forms", "anchorForms"]));
  const { t } = useTranslation(namespace);

  if (!error || !name) return null;

  const message = lookupErrorMessage(
    error,
    scope,
    name,
    inputType,
    t,
    namespace
  );

  return (
    <ErrorMessage
      errors={message.split("\n")}
      parentId={inputId}
      {...elementProps}
    />
  );
};
