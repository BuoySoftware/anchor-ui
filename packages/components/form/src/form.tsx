import kebabCase from "lodash/kebabCase";
import { useEffect, useState } from "react";
import {
  FieldValues,
  FormProvider,
  get,
  SubmitHandler,
  UseFormReturn,
} from "react-hook-form";
import { ErrorMessage } from "./error_message";
import { Box, BoxProps } from "@buoysoftware/anchor-layout";

import { FormScopeProvider, TNamespace } from "./form_scope_provider";

export type FormValidationErrors<T> = {
  [key in keyof T]: string[];
};

export type ErrorKeys<TFormValues> = Record<string, keyof TFormValues>;

interface OwnProps<TFormValues extends FieldValues> {
  "data-testid"?: string;
  id: string;
  scope: string;
  onSubmit?: SubmitHandler<TFormValues>;
  children: React.ReactNode;
  form: UseFormReturn<TFormValues>;
  errorKeys?: ErrorKeys<TFormValues>;
  mutationErrors?: Partial<FormValidationErrors<TFormValues>>;
  tNamespace?: TNamespace;
}

export type FormProps<TFormValues extends FieldValues> = OwnProps<TFormValues> &
  BoxProps;

export const Form = <TFormValues extends FieldValues>({
  "data-testid": testId,
  id,
  scope,
  children,
  errorKeys = {},
  form: { handleSubmit, ...form },
  onSubmit = () => undefined,
  mutationErrors = {},
  tNamespace,
  ...elementProps
}: FormProps<TFormValues>): React.ReactElement => {
  const [hasErrors, setHasErrors] = useState(false);

  useEffect(() => {
    if (hasErrors) {
      Object.entries(mutationErrors).forEach(([key, errors]) => {
        if (key !== "base") {
          const messages = errors?.join("\n");
          const inputKey = get(errorKeys, key, key);

          form.setError(inputKey as any, {
            type: "server",
            message: messages,
          });
        }
      });
    }
  }, [hasErrors, mutationErrors]);

  useEffect(() => {
    setHasErrors(!!Object.keys(mutationErrors).length);
  }, [mutationErrors]);

  const baseErrors = mutationErrors["base"] || [];

  return (
    <FormScopeProvider value={{ formId: id, baseErrors, scope, tNamespace }}>
      <FormProvider {...form} handleSubmit={handleSubmit}>
        <Box
          as="form"
          autoComplete="off"
          id={id}
          data-testid={testId || kebabCase(scope)}
          onSubmit={handleSubmit(onSubmit)}
          {...elementProps}
        >
          {children}
        </Box>
      </FormProvider>
    </FormScopeProvider>
  );
};
