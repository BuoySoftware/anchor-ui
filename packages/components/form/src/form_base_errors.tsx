import { useFormScope } from "./form_scope_provider";
import { ErrorMessage } from "./error_message";

export const FormBaseErrors: React.FC = (): React.ReactElement | null => {
  const { baseErrors, formId } = useFormScope();

  if (baseErrors.length === 0) return null;

  return (
    <ErrorMessage errors={baseErrors} parentId={formId} mb="form.fieldGap" />
  );
}
