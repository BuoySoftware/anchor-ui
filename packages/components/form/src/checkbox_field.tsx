import { Box, BoxProps } from "@buoysoftware/anchor-layout";
import { RegisterOptions, useFormContext } from "react-hook-form";

import { useFormField } from "./use_form_field";
import { InlineLabel } from "./inline_label";
import { Checkbox } from "./checkbox";
import { FieldError } from "./field_error";

interface OwnProps {
  name: string;
  value?: string;
  rules?: RegisterOptions;
}

type CheckboxFieldProps = OwnProps & BoxProps;

export type CheckboxFieldValue = "1" | false;

export const DEFAULT_CHECKED_VALUE: CheckboxFieldValue = "1";

export const CheckboxField: React.FC<CheckboxFieldProps> = ({
  name,
  value = DEFAULT_CHECKED_VALUE,
  rules,
  ...boxProps
}): React.ReactElement => {
  const { label, error, inputId } = useFormField({ name });
  const { register } = useFormContext();

  return (
    <Box
      as="fieldset"
      data-testid={`checkbox-field-${inputId}`}
      display="flex"
      flexDirection="column"
      mb="form.fieldGap"
      {...boxProps}
    >
      <InlineLabel
        data-testid={`label-${inputId}`}
        display="flex"
        alignItems="center"
        htmlFor={inputId}
      >
        <Checkbox
          aria-describedby={error ? `error-${name}` : undefined}
          aria-invalid={error ? "true" : "false"}
          data-testid={inputId}
          id={inputId}
          mr="xxs"
          value={value}
          {...register(name, { shouldUnregister: true, ...rules })}
        />
        {label}
      </InlineLabel>
      <FieldError
        error={error}
        inputId={inputId}
        inputType="checkbox"
        mt="s"
        name={name}
      />
    </Box>
  );
};
