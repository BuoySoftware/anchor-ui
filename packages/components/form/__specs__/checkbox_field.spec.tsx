import { useForm } from "react-hook-form";
import i18n from "i18next";
import { initReactI18next } from "react-i18next";

import { render as baseRender, waitFor, UserRenderResult } from "../test_utils";
import { CheckboxField } from "../src/checkbox_field";
import { Form } from "../src/form";

describe("<CheckboxField />", () => {
  const translations = {
    errors: {
      input_types: {
        checkbox: {
          required: "This field is required",
        },
      },
    },
    labels: {
      my_checkbox_field: "Example checkbox",
    },
  };

  i18n.use(initReactI18next).init({
    lng: "en",
    ns: ["forms"],
    resources: {
      en: {
        forms: translations,
      },
    },
  });

  const MockForm: React.FC = (): React.ReactElement => {
    const form = useForm();

    return (
      <Form
        id="test-form"
        data-testid="test-form"
        form={form}
        scope="test_form"
        onSubmit={jest.fn()}
      >
        <CheckboxField name="myCheckboxField" rules={{ required: true }} />
        <button type="submit">Submit</button>
      </Form>
    );
  };

  function render(): UserRenderResult {
    return baseRender(<MockForm />);
  }

  it("renders properly", () => {
    const { asFragment } = render();

    expect(asFragment()).toMatchSnapshot();
  });

  context("with error", () => {
    it("flags the field as invalid", async () => {
      const { user, getByText, getByLabelText } = render();

      await user.click(getByText("Example checkbox"));

      waitFor(() => {
        expect(getByLabelText("Example checkbox")).toBeInvalid();
      });
    });

    it("renders properly", async () => {
      const { user, getByText, asFragment } = render();

      await user.click(getByText("Submit"));

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
