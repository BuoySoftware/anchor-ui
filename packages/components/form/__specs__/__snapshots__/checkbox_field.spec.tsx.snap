// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxField /> renders properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c1 {
  box-sizing: border-box;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 1.5rem;
}

.c3 {
  box-sizing: border-box;
  color: #101112;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: m;
  height: m;
}

.c4 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-style: none;
  font-weight: 400;
}

[disabled] .c2 {
  color: #9CA1A5;
}

.c5 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: none;
  border: 2px solid #7F8387;
  border-radius: 4px;
  width: 18px;
  height: 18px;
  padding: 0.25rem;
  cursor: pointer;
  box-sizing: border-box;
  margin-right: 0.25rem;
}

.c5:hover {
  border-color: #46494D;
}

.c5:focus-visible {
  box-shadow: 0px 0px 0px 2px #FFFFFF, 0px 0px 0px 4px #5B91F4;
}

.c5:checked {
  border-color: transparent;
  background: center #276EF1 url(test-file-stub) no-repeat;
}

.c5:hover:checked {
  background-color: #1E54B7;
}

.c5:disabled {
  border-color: #BABEC3;
}

.c5:checked:disabled {
  background-color: #BABEC3;
}

<form
    autocomplete="off"
    class="c0"
    data-testid="test-form"
    id="test-form"
  >
    <fieldset
      class="c1"
      data-testid="checkbox-field-my-checkbox-field"
      display="flex"
    >
      <label
        class="c2 c3 c4 "
        color="text.primary"
        data-testid="label-my-checkbox-field"
        display="flex"
        font-style="none"
        font-weight="400"
        for="my-checkbox-field"
      >
        <input
          aria-invalid="false"
          class="c5"
          data-testid="my-checkbox-field"
          id="my-checkbox-field"
          name="myCheckboxField"
          type="checkbox"
          value="1"
        />
        Example checkbox
      </label>
    </fieldset>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`<CheckboxField /> with error renders properly 1`] = `
<DocumentFragment>
  .c0 {
  box-sizing: border-box;
}

.c1 {
  box-sizing: border-box;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 1.5rem;
}

.c3 {
  box-sizing: border-box;
  color: #101112;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: m;
  height: m;
}

.c6 {
  box-sizing: border-box;
  border-radius: 4px;
  background-color: #FFDED9;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-top: 0.25rem;
  padding-bottom: 0.25rem;
  padding-left: 0.5rem;
  padding-right: 0.5rem;
  margin-top: 0.75rem;
}

.c7 {
  box-sizing: border-box;
  margin-left: 0.5rem;
}

.c8 {
  box-sizing: border-box;
  color: #BF2F1D;
  width: m;
  height: m;
}

.c10 {
  box-sizing: border-box;
  color: inherit;
}

.c4 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-style: none;
  font-weight: 400;
}

.c9 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  font-style: none;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1rem;
}

.c11 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Segoe UI, Roboto,Helvetica Neue,sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

[disabled] .c2 {
  color: #9CA1A5;
}

.c5 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: none;
  border: 2px solid #7F8387;
  border-radius: 4px;
  width: 18px;
  height: 18px;
  padding: 0.25rem;
  cursor: pointer;
  box-sizing: border-box;
  margin-right: 0.25rem;
}

.c5:hover {
  border-color: #46494D;
}

.c5:focus-visible {
  box-shadow: 0px 0px 0px 2px #FFFFFF, 0px 0px 0px 4px #5B91F4;
}

.c5:checked {
  border-color: transparent;
  background: center #276EF1 url(test-file-stub) no-repeat;
}

.c5:hover:checked {
  background-color: #1E54B7;
}

.c5:disabled {
  border-color: #BABEC3;
}

.c5:checked:disabled {
  background-color: #BABEC3;
}

<form
    autocomplete="off"
    class="c0"
    data-testid="test-form"
    id="test-form"
  >
    <fieldset
      class="c1"
      data-testid="checkbox-field-my-checkbox-field"
      display="flex"
    >
      <label
        class="c2 c3 c4 "
        color="text.primary"
        data-testid="label-my-checkbox-field"
        display="flex"
        font-style="none"
        font-weight="400"
        for="my-checkbox-field"
      >
        <input
          aria-describedby="error-myCheckboxField"
          aria-invalid="true"
          class="c5"
          data-testid="my-checkbox-field"
          id="my-checkbox-field"
          name="myCheckboxField"
          type="checkbox"
          value="1"
        />
        Example checkbox
      </label>
      <div
        class="c6 "
        data-testid="error-my-checkbox-field"
        display="flex"
        id="error-my-checkbox-field"
        role="alert"
      >
        <svg
          color="#BF2F1D"
          fill="none"
          height="16"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 7v6M12 17.01l.01-.011M12 22c5.523 0 10-4.477 10-10S17.523 2 12 2 2 6.477 2 12s4.477 10 10 10z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        <div
          class="c7"
        >
          <div
            class="c8 c9"
            color="text.critical"
            font-size="font-size-100"
            font-style="none"
            font-weight="400"
          >
            <p
              class="c10 c11"
              color="inherit"
            >
              This field is required
            </p>
          </div>
        </div>
      </div>
    </fieldset>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
